pkg/lib/queueinformer/queueinformer.go

```diff
@@ -41,90 +76,116 @@ func (q *QueueInformer) enqueue(obj interface{}) {
+	return &cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if err := q.HandleMetrics(); err != nil {
+				q.logger.WithError(err).WithField("key", obj).Warn("error handling metrics on add event")
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			if err := q.HandleMetrics(); err != nil {
+				q.logger.WithError(err).WithField("key", obj).Warn("error handling metrics on delete event")
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if err := q.HandleMetrics(); err != nil {
+				q.logger.WithError(err).WithField("key", newObj).Warn("error handling metrics on update event")
+			}
+		},
 	}
-
-	q.queue.Forget(key)
 }
 
-func (q *QueueInformer) defaultUpdateFunc(oldObj, newObj interface{}) {
-	key, ok := q.keyFunc(newObj)
-	if !ok {
-		q.log.Warnf("couldn't update %v, couldn't create key", newObj)
-		return
+// NewQueueInformer returns a new QueueInformer configured with options.
+func NewQueueInformer(name string, queue workqueue.RateLimitingInterface, informer cache.SharedIndexInformer, options ...Option) *QueueInformer {
+	// Set default QueueInformer configuration
+	queueInformer := &QueueInformer{
+		MetricsProvider:     metrics.NewMetricsNil(),
+		SharedIndexInformer: informer,
+
+		name:   name,
+		queue:  queue,
+		logger: logrus.New(),
 	}
+	queueInformer.resourceHandlers = queueInformer.defaultResourceHandlers()
+	queueInformer.metricHandlers = queueInformer.defaultMetricHandlers()
+	queueInformer.AddSyncHandlers(queueInformer.metaLoggingSyncHandler)
 
-	q.enqueue(key)
-}
+	// Apply configuration options
+	queueInformer.apply(options...)
 
-// defaultResourceEventhandlerFuncs provides the default implementation for responding to events
-// these simply Log the event and add the object's key to the queue for later processing
-func (q *QueueInformer) defaultResourceEventHandlerFuncs() *cache.ResourceEventHandlerFuncs {
-	return &cache.ResourceEventHandlerFuncs{
-		AddFunc:    q.defaultAddFunc,
-		DeleteFunc: q.defaultDeleteFunc,
-		UpdateFunc: q.defaultUpdateFunc,
-	}
+	queueInformer.AddEventHandler(queueInformer.resourceHandlers)
+	queueInformer.AddEventHandler(queueInformer.metricHandlers)
+
+	return queueInformer
 }
 
-// New creates a set of new queueinformers given a name, a set of informers, and a sync handler to handle the objects
+// NewQueueInformers creates a set of new queueinformers given a name, a set of informers, and a sync handler to handle the objects
 // that the operator is managing. Optionally, custom event handler funcs can be passed in (defaults will be provided)
-func New(queue workqueue.RateLimitingInterface, informers []cache.SharedIndexInformer, handler SyncHandler, funcs *cache.ResourceEventHandlerFuncs, name string, metrics metrics.MetricsProvider, logger *logrus.Logger) []*QueueInformer {
+func NewQueueInformers(name string, queue workqueue.RateLimitingInterface, informers []cache.SharedIndexInformer, options ...Option) []*QueueInformer {
```

Comment:
I felt that `Builder`s composed better than options; i.e. it's hard to write a constructor similar to the one above for a type that embeds `QueueInformer`, since options work on parameters of a concrete type and the variadic parameter must be the last one in the function signature. To do that, I think we would need to create a wrapper function of the embedding type's option type; e.g.:

```go
type EmbeddedOption func(embedded *Embedded)
/// ... Some options for the embedded type
type Embedding struct {
  *Embedded
}
// ...
type EmbeddingOption func(embedding *Embedding)
// ... Some embedding options
func WithEmbeddedOption(options ...EmbeddedOption) EmbeddingOption {
  return func(embedding *Embedding) {
    for _, option := range options {
      option(embedding.embedded)
    }
  }
}
```

There could be something obvious I'm missing here though. Any suggestions are definitely welcome.